src/java.base/share/classes/jdk/internal/reflect/DirectConstructorAccessorImpl.java

```diff
@@ -0,0 +1,157 @@
+import jdk.internal.vm.annotation.Stable;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.WrongMethodTypeException;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+
+import static jdk.internal.reflect.MethodHandleAccessorFactory.SPECIALIZED_PARAM_COUNT;
+import static jdk.internal.reflect.MethodHandleAccessorFactory.newMethodHandleInvoker;
+
+class DirectConstructorAccessorImpl extends ConstructorAccessorImpl {
+    static ConstructorAccessorImpl constructorAccessor(Constructor<?> ctor, MethodHandle target) {
+        if (ReflectionFactory.noInflation()) {
+            // fast invoker
+            var mhInvoker = newMethodHandleInvoker(ctor, target);
+            return new DirectConstructorAccessorImpl(ctor, target);
+        } else {
+            // Default is the adaptive accessor method.
+            return new AdaptiveConstructorAccessor(ctor, target);
+        }
+    }
+
+    static ConstructorAccessorImpl nativeAccessor(Constructor<?> ctor) {
+        return new NativeAccessor(ctor);
+    }
+
+    protected final Constructor<?> ctor;
+    protected final int paramCount;
+
+    @Stable protected final MethodHandle target;
+    @Stable protected final MHInvoker invoker;
+    DirectConstructorAccessorImpl(Constructor<?> ctor, MethodHandle target) {
+        this.ctor = ctor;
+        this.paramCount = ctor.getParameterCount();
+        this.target = target;
+        this.invoker = new MHInvokerDelegate(target);
+    }
+
+    @ForceInline
+    MHInvoker mhInvoker() {
+        return invoker;
+    }
+
+    @Override
+    public Object newInstance(Object[] args) throws InstantiationException, InvocationTargetException {
+        int argc = args != null ? args.length : 0;
+        // only check argument count for specialized forms
+        if (paramCount <= SPECIALIZED_PARAM_COUNT && argc != paramCount) {
+            throw new IllegalArgumentException("wrong number of arguments: " + argc + " expected: " + paramCount);
+        }
+        try {
+            return invokeImpl(args);
+        } catch (ClassCastException|WrongMethodTypeException e) {
+            if (isIllegalArgument(e))
+                throw new IllegalArgumentException("argument type mismatch", e);
+            else
+                throw new InvocationTargetException(e);
+        } catch (NullPointerException e) {
+            if (isIllegalArgument(e))
+                throw new IllegalArgumentException(e);
+            else
+                throw new InvocationTargetException(e);
+        } catch (Throwable e) {
+            throw new InvocationTargetException(e);
+        }
+    }
+
+    private boolean isIllegalArgument(RuntimeException ex) {
+        return AccessorUtils.isIllegalArgument(DirectConstructorAccessorImpl.class, ex);
+    }
+
+    @Hidden
+    @ForceInline
+    Object invokeImpl(Object[] args) throws Throwable {
+        var mhInvoker = mhInvoker();
+        return switch (paramCount) {
```

Comment:
While I recall it was motivated primarily for startup (note startup numbers for `-Djdk.reflect.useSpreader=true` in @mlchung  reply earlier in this PR), the specialization to avoid the spreader for low-arity arguments also improve performance on throughput microbenchmarks. Surprisingly also reduce the per invocation allocation rate. 

Allocation profiling suggests that with the specialization a varargs array is entirely eliminated. This is of course somewhat fragile and dependent on a number of things - and might ultimately be an artifact of a rather synthetic microbenchmark that will have little benefit in practice, but it's a rather consistent gain for various number of arguments - even when actually going into a spreader (I have some hypotheses as to why this might happen, but most likely we help the JIT to narrow things down for each kind of invocation scheme with the selector):
```
18 baseline
Benchmark                                                                  Mode  Cnt    Score     Error   Units
ReflectionMethods.static_method                                            avgt   10   57.329 ±   4.217   ns/op
ReflectionMethods.static_method:·gc.alloc.rate.norm                        avgt   10   72.006 ±   0.001    B/op
ReflectionMethods.static_method_3arg                                       avgt   10   70.940 ±   7.457   ns/op
ReflectionMethods.static_method_3arg:·gc.alloc.rate.norm                   avgt   10   96.008 ±   0.002    B/op
ReflectionMethods.static_method_4arg                                       avgt   10   90.453 ±   4.252   ns/op
ReflectionMethods.static_method_4arg:·gc.alloc.rate.norm                   avgt   10  112.010 ±   0.001    B/op

pull/5027
Benchmark                                                                  Mode  Cnt    Score     Error   Units
ReflectionMethods.static_method                                            avgt   10   41.518 ±   2.444   ns/op
ReflectionMethods.static_method:·gc.alloc.rate.norm                        avgt   10   48.004 ±   0.001    B/op
ReflectionMethods.static_method_3arg                                       avgt   10   57.603 ±   3.299   ns/op
ReflectionMethods.static_method_3arg:·gc.alloc.rate.norm                   avgt   10   64.006 ±   0.001    B/op
ReflectionMethods.static_method_4arg                                       avgt   10   56.772 ±   3.971   ns/op
ReflectionMethods.static_method_4arg:·gc.alloc.rate.norm                   avgt   10   80.007 ±   0.001    B/op
```
On a patch to remove the specialization on top of pull/5027, performance reverts back to numbers very similar to the baseline:
```
remove-spreader-patch:
Benchmark                                                                  Mode  Cnt    Score     Error   Units
ReflectionMethods.static_method                                            avgt   10   56.644 ±   4.322   ns/op
ReflectionMethods.static_method:·gc.alloc.rate.norm                        avgt   10   72.006 ±   0.001    B/op
ReflectionMethods.static_method_3arg                                       avgt   10   72.353 ±   6.815   ns/op
ReflectionMethods.static_method_3arg:·gc.alloc.rate.norm                   avgt   10   96.008 ±   0.001    B/op
ReflectionMethods.static_method_4arg                                       avgt   10   82.820 ±   8.303   ns/op
ReflectionMethods.static_method_4arg:·gc.alloc.rate.norm                   avgt   10  112.009 ±   0.002    B/op
```

I think the cold start reduction alone is worth the extra 100 lines or so of code, but if the gain on these microbenchmarks translates to real throughput gains then I think the complexity is more than paid for. Simplifying it while retaining similar characteristics would be great of course, but I think such an exploration should be done as a follow-up.